pkg/utils/utils.go

```diff
@@ -197,8 +198,71 @@ func DiscoverSriovDevices(withUnsupported bool, storeManager StoreManagerInterfa
 }
 
 // SyncNodeState Attempt to update the node state to match the desired state
-func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool) error {
-	return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool, parallelNicConfig bool) error {
+	if !parallelNicConfig {
+		return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+	}
+	return ConfigSriovInterfacesInParallel(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+}
+
+func ConfigSriovInterfacesInParallel(interfaces []sriovnetworkv1.Interface, ifaceStatuses []sriovnetworkv1.InterfaceExt, pfsToConfig map[string]bool) error {
+	glog.V(2).Infof("ConfigSriovInterfacesInParallel(): start sriov configuration")
+	if IsKernelLockdownMode(true) && hasMellanoxInterfacesInSpec(ifaceStatuses, interfaces) {
+		glog.Warningf("cannot use mellanox devices when in kernel lockdown mode")
+		return fmt.Errorf("cannot use mellanox devices when in kernel lockdown mode")
+	}
+	// TODO(e0ne): store all errors in SriovNetworkNodeState
+	var result error
+	wg := sync.WaitGroup{}
+	for _, ifaceStatus := range ifaceStatuses {
+		configured := false
+		for _, iface := range interfaces {
+			if iface.PciAddress == ifaceStatus.PciAddress {
+				configured = true
+
+				if skip := pfsToConfig[iface.PciAddress]; skip {
+					break
+				}
+
+				if !NeedUpdate(&iface, &ifaceStatus) {
+					glog.V(2).Infof("syncNodeState(): no need update interface %s", iface.PciAddress)
+					break
+				}
+
+				wg.Add(1)
+				go func(iface *sriovnetworkv1.Interface, ifaceStatus *sriovnetworkv1.InterfaceExt) {
+					if err := configSriovDevice(iface, ifaceStatus); err != nil {
+						glog.Errorf("ConfigSriovInterfacesInParallel(): fail to configure sriov interface %s: %v. resetting interface.", iface.PciAddress, err)
+						result = err
```

Comment:
What about using [`errgroup.Group`](https://cs.opensource.google/go/x/sync/+/v0.4.0:errgroup/errgroup.go;l=22) to handle this synchronization? I think it would be useful, as we are dealing with errors here.

Reply:
we need to re-visit our API to include muliple results in SriovNetworkNodeState